SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp

```diff
@@ -437,22 +423,148 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
 	return HandoverChanged;
 }
 
-void USpatialActorChannel::GetLatestAuthorityChangeFromHierarchy(const AActor* HierarchyActor, uint64& OutTimestamp)
+template <typename Functor>
+TFunction<void(AActor*)> RebindActorChannel(USpatialNetDriver* NetDriver, Functor&& InFunctor)
+{
+	return [NetDriver, InFunctor](AActor* Actor)
+	{
+		if (USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(Actor))
+		{
+			InFunctor(Actor, Channel);
+		}
+	};
+}
+
+void USpatialActorChannel::GetLatestAuthorityChangeFromHierarchy(USpatialNetDriver* NetDriver, const AActor* RootActor, uint64& OutTimestamp)
 {
-	if (HierarchyActor->GetIsReplicated())
+	SpatialGDK::ForeachReplicatedActorInHierarchy(const_cast<AActor*>(RootActor), RebindActorChannel(NetDriver, [&OutTimestamp] (AActor*, USpatialActorChannel* Channel)
 	{
-		if (USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(const_cast<AActor*>(HierarchyActor)))
+		if (Channel->AuthorityReceivedTimestamp > OutTimestamp)
 		{
-			if (Channel->AuthorityReceivedTimestamp > OutTimestamp)
+			OutTimestamp = Channel->AuthorityReceivedTimestamp;
+		}
+	}));
+}
+
+void USpatialActorChannel::MarkActorHierarchyForMigration(USpatialNetDriver* NetDriver, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [NetDriver](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			ChildActor->ForceNetUpdate();
+			Channel->MigrationVirtualWorkerId.Emplace(SpatialConstants::INVALID_VIRTUAL_WORKER_ID);
+			Channel->MigrationInitiatedFrame = NetDriver->ReplicationFrame;
+		}
+	}));
+}
+
+void USpatialActorChannel::SetActorHierarchyMigrationDestination(USpatialNetDriver* NetDriver, VirtualWorkerId NewAuthVirtualWorkerId, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [NewAuthVirtualWorkerId] (AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if(ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId = NewAuthVirtualWorkerId;
+		}
+	}));
+}
+
+void USpatialActorChannel::DropActorHierarchyMigration(USpatialNetDriver* NetDriver, AActor* RootActor)
+{
+	SpatialGDK::ForeachReplicatedActorInHierarchy(RootActor, RebindActorChannel(NetDriver, [](AActor* ChildActor, USpatialActorChannel* Channel)
+	{
+		if (ChildActor->HasAuthority())
+		{
+			Channel->MigrationVirtualWorkerId.Reset();
+		}
+	}));
+}
+
+void USpatialActorChannel::HandleActorLoadBalancing()
+{
+	// if the field is set it means we should try to migrate authority to another worker
+	if (MigrationVirtualWorkerId.IsSet())
```

Comment:
I think you can replace `MigrationWorkerVirtualId.IsSet()` with `MigrationFrame != NetDriver->ReplicationFrame` and make MigrationWorkerVirtualId not be an optional anymore?

Reply:
Replied in the previous comment-->

They are not the same state.
MigrationWorkerVirtualId.IsSet() means that someone in the hierarchy has requested a migration on this frame, or on a previous frame.
MigrationFrame != NetDriver->ReplicationFrame means that we are at a different replication frame than the one where we determined we should migrate, and at this frame, all actors of the hierarchy will go through replication.